lib/baby_squeel/table.rb

```diff
@@ -84,7 +85,13 @@ def evaluate(&block)
     # This method allows BabySqueel::Nodes::Attribute
     # instances to find what their alias will be.
     def find_alias(associations = [])
-      rel = _scope.joins _arel(associations)
+      this_arel = _arel(associations)
+      if this_arel.kind_of?(Hash)
+        _accumulate_associations(@_accumulated_associations, this_arel)
+        rel = _scope.joins @_accumulated_associations
+      else
+        rel = _scope.joins this_arel
+      end
```

Comment:
Can you tell me more about what's happening here? Does `@_accumulated_associations` need to be an instance variable?

Also, all of BabySqueel's nodes should be immutable to facilitate chaining.

Reply:
The aliases assigned to tables by Active Record depend on the entire set of associations incorporated in the whole query.  Therefore, the whole query, rooted at the `BabySqueel::Table`, needs to track the full set of associations used or there is no way to get the correct alias names.

I added `@_accumulated_associations` strictly for this purpose.  It is initialized in `BabySqueel::Table#initialize` to an empty `Hash` and is updated whenever associations are built off that instance of `BabySqueel::Table`.  Would it make sense to customize `#initialize_copy` to make a deep clone of `@_accumulated_associations`?

Reply:
That sounds like a possible solution for the immutability aspect. But, I guess I don't fully understand why `_accumulated_associations` is necessary? What changed that made this approach necessary? Because, in previous versions of this package, we were able to track the aliases.

Reply:
Alternatively, I could implement an immutable tree (similar to [this](https://voormedia.com/blog/2013/02/creating-immutable-tree-data-structures-in-ruby)), which would only need a shallow clone during `#initialize_copy`.  This could also, perhaps, be tracked on the `#_scope` (again with shallow duplication in `#initialize_copy`) so it is available to chained methods.

Additional alternatives:
* put the accumulated associations into the `#_scope` and extend `#initialize_copy` as a

Reply:
It looks like version 5.2.1 of Active Record changed the way aliases are assigned to tables.  I tested my expression `Post.joining {[author, comments.author]}.where.has {(author.name == 'Alice') & (comments.author.name == 'Bob')}.to_sql` against Active Record 5.2.0 and it correctly tested the `name` column of the aliased use of `authors` (`authors_comments`) for equality to `'Bob'`, where with Active Record 2.5.1, the generated SQL _incorrectly_ tests `"authors"."name" = 'Bob'`.  I tested both versions of Active Record against commit a6a9f43 of BabySqueel.

Reply:
And I've just re-tested with commit eba126a (before adding `@_accumulated_associations`) and I'm now seeing it generating the expressions in the `WHERE` clause.  I'll revert commit 90e02e2.